### DISCUSSION TITLE
How to prevent the following error: TypeError: Cannot read properties of undefined (reading 'constructor')

### DISCUSSION THREAD

**Discussion by simathedev** [sidorares/node-mysql2#1700] [Q&A] [answered]
Hello, could you please assist me. I don't know what caused this error and how I can prevent it from happening again.

I'm using mysql 2 in my telegraf bot to perform CRUD operations.

One unregistered user bypassed the system. I'm not sure if it's because the SQL statement did not run in this instance becuase
there are measures to prevent unregistered users from using the bot.

**The login process:**

1.  User enters their phone number.
2. The phone number is looked up in tblusers in the database using a select statement.
3. If the phone number is not found the user is denied access.

The unregistered user was able to use the bot until they encountered an error when the bot tried to insert data in another table and the UserID was null.

**The terminal displayed the following error: **
`TypeError: Cannot read properties of undefined (reading 'constructor')`

**Where the error took place:**
```
 query(sql, values, cb) {
    let cmdQuery;
    if (sql.constructor === Commands.Query) {
      cmdQuery = sql;
    } else {
      cmdQuery = Connection.createQuery(sql, values, cb, this.config);
    }
    this._resolveNamedPlaceholders(cmdQuery);
    const rawSql = this.format(cmdQuery.sql, cmdQuery.values !== undefined ? cmdQuery.values : []);
    cmdQuery.sql = rawSql;
    return this.addCommand(cmdQuery);
```
  }

**Comment by sidorares** [comment, marked as answer]
looks like the sql passed to `.query()` is undefined ( we expect a string here ). Can yuu double check how this could happen? ( uninitialised variable etc )

**Reply by simathedev** [reply]
It was an uninitialized variable but I'm honestly not sure how this is possible because I make use of if statements for data validation. very odd.
